File: 1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one/main.go
package main

import (
	"fmt"
	"math/big"
)

func numSteps(s string) int {

	n := new(big.Int)
	n, success := n.SetString(s, 2)
	if !success {
		fmt.Println("error parsing the binary string")
		return 0
	}

	var countSteps int = 0
	one := big.NewInt(1)
	two := big.NewInt(2)

	for n.Cmp(one) != 0 {
		if new(big.Int).And(n, one).Cmp(one) == 0 {
			n.Add(n, one)
		} else {
			n.Div(n, two)
		}
		countSteps++
	}

	return countSteps
}

File: 1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one/main.py
class Solution:
    def numSteps(self, s: str) -> int:

        n = int(s, 2)

        count_steps = 0
        while n != 1:
            if n & 1 != 0:
                n += 1
            else:
                n //= 2
                
            count_steps+=1
        
        return count_steps
        
